platforms/core-runtime/service-provider/src/main/java/org/gradle/internal/service/ServiceRegistration.java

```diff
@@ -40,7 +41,18 @@ public interface ServiceRegistration {
      * @param serviceType The service to make visible.
      * @param implementationType The implementation type of the service.
      */
-    void add(Class<?> serviceType, Class<?> implementationType);
+    <S, I extends S> void add(Class<S> serviceType, Class<I> implementationType);
+
+    /**
+     * Adds two services to this registry that share the implementation.
+     * <p>
+     * The implementation class should have a single public constructor, and this constructor can take services to be injected as parameters.
+     *
+     * @param serviceType1 The first service to make visible.
+     * @param serviceType2 The second service to make visible.
+     * @param implementationType The implementation type of the service.
+     */
+    <S, I extends S> void add(Class<S> serviceType1, Class<?> serviceType2, Class<I> implementationType);
```

Comment:
💭  This is the main new API for registering double-service implementations.

The second service type is not bound by a generic, because Java does not support `I extends S1 & S2` construct, when `S2` is not a specific interface. Having said that, the generics are mainly the ergonomics for the API users to get an earlier feedback about a mistake. At runtime, we validate that the implementation class actually implements all declared services.

Reply:
It seems it does! Good idea 👍 
I'll address this in a follow up